src/commands/commands.ts

```diff
@@ -1,29 +1,122 @@
 import {Command, flags} from '@oclif/command'
-import ux from 'cli-ux'
+import {ux} from 'cli-ux'
 import * as _ from 'lodash'
+import {EOL} from 'os'
 
+type Dictionary = {[index: string]: object}
 export default class Commands extends Command {
   static description = 'list all the commands'
 
-  static flags = {
+  static flags: flags.Input<any> = {
     help: flags.help({char: 'h'}),
-    json: flags.boolean({char: 'j', description: 'output in json format'}),
-    hidden: flags.boolean({description: 'also show hidden commands'}),
+    json: flags.boolean({char: 'j', description: 'display unfiltered api data in json format'}),
+    hidden: flags.boolean({description: 'show hidden commands'}),
+    ...ux.table.flags(),
   }
 
   async run() {
     const {flags} = this.parse(Commands)
-    let commands = this.config.commands
+    let commands = this.getCommands()
     if (!flags.hidden) {
       commands = commands.filter(c => !c.hidden)
     }
-    commands = _.sortBy(commands, 'id')
+
+    const config = this.config
+    commands = _.sortBy(commands, 'id').map(command => {
+      // Template supported fields.
+      command.description = (typeof command.description === 'string' && _.template(command.description)({command, config})) || undefined
+      command.usage = (typeof command.usage === 'string' && _.template(command.usage)({command, config})) || undefined
+      return command
+    })
+
     if (flags.json) {
-      ux.styledJSON(commands)
+      ux.styledJSON(commands.map(cmd => {
+        let commandClass = cmd.load()
+        const obj = Object.assign({}, cmd, commandClass)
+
+        // Load all properties on all extending classes.
+        while (commandClass !== undefined) {
+          commandClass = Object.getPrototypeOf(commandClass) || undefined
+          Object.assign(obj, commandClass)
+        }
+
+        // The plugin property on the loaded class contains a LOT of information including all the commands again. Remove it.
+        delete obj.plugin
+
+        // If Command classes have circular references, don't break the commands command.
+        return this.removeCycles(obj)
+      }))
     } else {
-      for (const c of commands) {
-        this.log(c.id)
+      ux.table(commands.map(command => {
+        // Message some fields so it looks good in the table
+        command.description = (command.description || '').split(EOL)[0]
+        command.hidden = Boolean(command.hidden)
+        command.usage = (command.usage || '')
+        return command
+      }), {
+        id: {
+          header: 'Command',
+        },
+        description: {},
+        usage: {
```

Comment:
what do you think of dropping usage?

Reply:
Usage is actually one of the main request from our doc writters.